src/lsm/grid.zig

```diff
@@ -19,54 +21,140 @@ pub fn GridType(comptime Storage: type) type {
+            return a == b;
+        }
+    };
+
+    const set_associative_cache_ways = 16;
+    const Cache = SetAssociativeCache(
+        u64,
+        [block_size]u8,
+        cache_interface.address_from_block,
+        cache_interface.hash_address,
+        cache_interface.equal_addresses,
+        .{
+            .ways = set_associative_cache_ways,
+            .value_alignment = config.sector_size,
+        },
+    );
+
+    const read_iops_max = 16;
+    assert(read_iops_max <= set_associative_cache_ways);
+
+    // TODO put more thought into how low/high this limit should be.
+    const write_iops_max = 16;
+
     const grid_offset: u64 = superblock_zone_size +
         write_ahead_log_zone_size +
         client_table_zone_size;
 
     return struct {
         const Grid = @This();
 
-        pub const Write = Storage.Write;
+        pub const Write = struct {
+            callback: fn (*Grid.Write) void,
+            address: u64,
+            block: BlockPtrConst,
 
-        pub const Read = struct {
+            /// Link for the writes_pending linked list.
+            next: ?*Write = null,
+
+            /// Call the user's callback, finishing the write.
+            pub fn finish(write: *Write) void {
+                const callback = write.callback;
+                write.* = undefined;
+                callback(write);
+            }
+        };
+
+        pub const WriteIOP = struct {
             grid: *Grid,
-            callback: fn (*Grid.Read) void,
-            completion: Storage.Read,
-            block: BlockPtr,
+            completion: Storage.Write,
+            write: *Write,
+        };
+
+        pub const Read = struct {
+            callback: fn (*Grid.Read, BlockPtrConst) void,
             address: u64,
             checksum: u128,
 
-            /// Link for read_recovery_queue linked list.
+            /// Link for reads_pending/read_recovery_queue/ReadIOP.reads linked lists.
             next: ?*Read = null,
 
             /// Call the user's callback, finishing the read.
             /// May be called by Replica after recovering the block over the network.
-            pub fn finish(read: *Read) void {
+            pub fn finish(read: *Read, block: BlockPtrConst) void {
                 const callback = read.callback;
                 read.* = undefined;
-                callback(read);
+                callback(read, block);
             }
         };
 
+        pub const ReadIOP = struct {
+            grid: *Grid,
+            completion: Storage.Read,
+            reads: FIFO(Read) = .{},
+            /// This is a pointer to a value in the block cache.
+            block: BlockPtr,
+        };
+
         superblock: *SuperBlock,
+        cache: Cache,
+
+        writes_pending: FIFO(Write) = .{},
+        write_iops: IOPS(WriteIOP, write_iops_max) = .{},
+
+        reads_pending: FIFO(Read) = .{},
+        read_iops: IOPS(ReadIOP, read_iops_max) = .{},
```

Comment:
```suggestion
        write_iops: IOPS(WriteIOP, write_iops_max) = .{},
        write_queue: FIFO(Write) = .{},

        read_iops: IOPS(ReadIOP, read_iops_max) = .{},
        read_queue: FIFO(Read) = .{},
```

To line up, and also to match the order that we use them: first try to acquire from `write_iops`, otherwise push to the queue.

Using a noun can also be very useful. For example, as an adjective, `pending` is good for something like a boolean state. But for a queue, a noun is better because that can more easily be referred to directly in external docs/communication, e.g. "Grid's write queue is full" without having to be rephrased e.g. "Grid has no pending writes".